libhoney.go

```diff
@@ -428,41 +445,87 @@ func SendNow(data interface{}) error {
 		return err
 	}
 	err := ev.Send()
-	logger.Printf("SendNow enqueued event, err=%v", err)
+	dc.logger.Printf("SendNow enqueued event, err=%v", err)
 	return err
 }
 
 // Responses returns the channel from which the caller can read the responses
 // to sent events.
 func Responses() chan Response {
-	return responses
+	return dc.responses
 }
 
 // AddDynamicField takes a field name and a function that will generate values
 // for that metric. The function is called once every time a NewEvent() is
 // created and added as a field (with name as the key) to the newly created
 // event.
 func AddDynamicField(name string, fn func() interface{}) error {
-	return defaultBuilder.AddDynamicField(name, fn)
+	ensureDefaultBuilder()
+	return dc.defaultBuilder.AddDynamicField(name, fn)
 }
 
 // AddField adds a Field to the global scope. This metric will be inherited by
 // all builders and events.
 func AddField(name string, val interface{}) {
-	defaultBuilder.AddField(name, val)
+	ensureDefaultBuilder()
+	dc.defaultBuilder.AddField(name, val)
 }
 
 // Add adds its data to the global scope. It adds all fields in a struct or all
 // keys in a map as individual Fields. These metrics will be inherited by all
 // builders and events.
 func Add(data interface{}) error {
-	return defaultBuilder.Add(data)
+	ensureDefaultBuilder()
+	return dc.defaultBuilder.Add(data)
 }
 
 // NewEvent creates a new event prepopulated with any Fields present in the
 // global scope.
 func NewEvent() *Event {
-	return defaultBuilder.NewEvent()
+	ensureDefaultBuilder()
+	startTx()
+	return dc.defaultBuilder.NewEvent()
+}
+
+// NewBuilder creates a new event builder. The builder inherits any
+// Dynamic or Static Fields present in the global scope.
+func NewBuilder() *Builder {
+	ensureDefaultBuilder()
+	startTx()
+	return dc.defaultBuilder.Clone()
+}
+
+func ensureDefaultBuilder() {
```

Comment:
Is this lazy initialization to prevent the application from experiencing errors if they used libhoney, but haven't called `Init()`? In the python sdk we lean towards just no-oping and printing a warning if uninitialized. The Tx/Builder here wouldn't work due to lack of a valid API key if Init hasn't been called.

Reply:
You can choose to explicitly set the API key etc. on a per-builder or per-event basis, so there is a path to skipping initialization. I suppose noop is an alternative, though there are a lot of nil references to protect if it's uninitialized.